File: c/socket-server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>
#include <netinet/in.h>

// Author: George Wen
// A C implementation for a multi-client socket chat program.
// This server manages messages between clients, allowing users to log in, compose messages, and read messages.

// Definition of a message structure in a linked list format.
typedef struct Message {
    char* from;           // Sender username
    char* text;           // Message text
    struct Message* next; // Pointer to the next message in the list
} Message;

// Dictionary entry structure, acting as a key-value pair where the key is the username.
typedef struct DictionaryEntry {
    char* key;                // Username as the key
    Message* messages;        // Pointer to the first message in the list for this user
    struct DictionaryEntry* next; // Pointer to the next dictionary entry in the list
} DictionaryEntry;

DictionaryEntry* dictionary = NULL; // Global dictionary to store messages for each user

// Function to insert a message for a specific user
void insertMessage(char* key, char* from, char* text) {
    // Iterate through the dictionary to find an entry for the key
    DictionaryEntry* current = dictionary;
    while (current != NULL) {
        if (strcmp(current->key, key) == 0) {
            // If the key exists, prepend a new message to the list
            Message* newMessage = (Message*)malloc(sizeof(Message));
            newMessage->from = strdup(from);
            newMessage->text = strdup(text);
            newMessage->next = current->messages;
            current->messages = newMessage;
            return;
        }
        current = current->next;
    }
    
    // If the key doesn't exist, create a new dictionary entry and prepend it
    DictionaryEntry* newEntry = (DictionaryEntry*)malloc(sizeof(DictionaryEntry));
    newEntry->key = strdup(key);
    newEntry->messages = NULL; // Initially, no messages
    newEntry->next = dictionary;
    dictionary = newEntry;
    
    // Add the new message to the newly created entry
    Message* newMessage = (Message*)malloc(sizeof(Message));
    newMessage->from = strdup(from);
    newMessage->text = strdup(text);
    newMessage->next = newEntry->messages;
    newEntry->messages = newMessage;
}

// Function to pop (retrieve and remove) the first message for a specific user
Message* popMessage(char* key) {
    DictionaryEntry* current = dictionary;
    while (current != NULL) {
        if (strcmp(current->key, key) == 0) {
            // If the user has messages, detach the first message from the list and return it
            Message* firstMessage = current->messages;
            if (firstMessage != NULL) {
                current->messages = firstMessage->next;
                return firstMessage;
            }
            break;
        }
        current = current->next;
    }
    // If no messages or the key doesn't exist, return NULL
    return NULL;
}

// Function to count the number of messages for a specific user
int countMessages(char* key) {
    DictionaryEntry* current = dictionary;
    while (current != NULL) {
        if (strcmp(current->key, key) == 0) {
            int count = 0;
            Message* messagePtr = current->messages;
            while (messagePtr != NULL) {
                count++;
                messagePtr = messagePtr->next;
            }
            return count; // Return the number of messages
        }
        current = current->next;
    }
    // If the key doesn't exist, return 0
    return 0;
}

#define BUFFER_SIZE 256 // Buffer size for reading messages
pthread_mutex_t clients_mutex = PTHREAD_MUTEX_INITIALIZER; // Mutex for thread-safe operations on the dictionary

// Thread function to handle communication with a client
void *handleClient(void *arg) {
    int client_socket = *((int *)arg);
    char buffer[BUFFER_SIZE];
    int readSize;

    char currentUser[50] = {0}; // Store the current user's name
    int receiving = 0;          // Flag to indicate if the server is expecting a message
    char toUser[50] = {0};      // Store the recipient user for messages

    while (1) {
        memset(buffer, 0, BUFFER_SIZE); // Clear the buffer
        readSize = recv(client_socket, buffer, BUFFER_SIZE, 0); // Read message from client

        if (readSize <= 0) {
            printf("exiting...\n");
            close(client_socket);
            break;
        }

        printf("Message received: %s\n", buffer);

        // Process commands
        if (strlen(currentUser) == 0) { // If the user has not logged in
            if (strncmp(buffer, "LOGIN", 5) == 0) {
                sscanf(buffer, "LOGIN %s", currentUser);
                int messageCount = countMessages(currentUser);
                char response[256];
                sprintf(response, "%d Messages\n", messageCount);
                send(client_socket, response, strlen(response), 0);
            } else {
                printf("Please issue LOGIN command first\n");
            }
        } else {
            if (receiving) {
                // Store the received message
                char* response = "MESSAGE SENT\n";
                pthread_mutex_lock(&clients_mutex);
                insertMessage(toUser, currentUser, buffer);
                pthread_mutex_unlock(&clients_mutex);
                send(client_socket, response, strlen(response), 0);
                receiving = 0;
            } else {
                // Handle READ, COMPOSE, or EXIT commands
                if (strncmp(buffer, "READ", 4) == 0) {
                    pthread_mutex_lock(&clients_mutex);
                    Message* msg = popMessage(currentUser);
                    pthread_mutex_unlock(&clients_mutex);
                    if (msg != NULL) {
                        send(client_socket, msg->from, strlen(msg->from), 0);
                        send(client_socket, "\n", 1, 0);
                        send(client_socket, msg->text, strlen(msg->text), 0);
                        send(client_socket, "\n", 1, 0);
                        free(msg->from); // Free the memory allocated for the message
                        free(msg->text);
                        free(msg);
                    } else {
                        char* response = "READ ERROR\n";
                        send(client_socket, response, strlen(response), 0);
                    }
                } else if (strncmp(buffer, "COMPOSE", 7) == 0) {
                    sscanf(buffer, "COMPOSE %s", toUser);
                    receiving = 1; // Next message will be stored as a message content
                } else if (strncmp(buffer, "EXIT", 4) == 0) {
                    break; // Exit the loop and close the socket
                } else {
                    char* response = "INVALID COMMAND\n";
                    send(client_socket, response, strlen(response), 0); // Inform the client of the invalid command
                }
            }
        }
    }

    printf("Exiting thread...\n");
    close(client_socket); // Close the client socket on exit
    // pthread_exit(NULL); // Optionally, explicitly exit the thread
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <port>\n", argv[0]);
        return 1;
    }

    int port = atoi(argv[1]); // Convert port argument from string to integer

    int serverSocket;
    struct sockaddr_in server_addr, clientAddr;
    pthread_t tid;

    // Create socket
    serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (serverSocket == -1) {
        perror("Could not create socket");
        return 1;
    }
    printf("Socket created\n");

    // Prepare the sockaddr_in structure
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = INADDR_ANY; // Accept connections on any IP address
    server_addr.sin_port = htons(port);       // Host to network short conversion for port

    // Bind
    if (bind(serverSocket, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("Bind failed");
        return 1;
    }
    printf("Bind done\n");

    // Listen for incoming connections
    listen(serverSocket, 3);
    printf("Waiting for incoming connections...\n");

    int c = sizeof(struct sockaddr_in);
    while (1) {
        int *clientSocket = malloc(sizeof(int));
        *clientSocket = accept(serverSocket, (struct sockaddr *)&clientAddr, (socklen_t*)&c);
        if (*clientSocket < 0) {
            perror("Accept failed");
            return 1;
        }
        printf("Connection accepted\n");

        // Create a new thread for each client
        if (pthread_create(&tid, NULL, handleClient, (void*)clientSocket) < 0) {
            perror("Could not create thread");
            return 1;
        }
    }

    return 0; // End of program
}

File: python/socket-server.py
# This is the chat server program written in Python.
# Author: George Wen
# Usage: Execute this script with a port number as an argument to start the server.

import socket
import threading
from collections import defaultdict
import sys

# Server IP address and port number are set here.
# '127.0.0.1' is the loopback address, meaning the server will only be accessible locally.
HOST = '127.0.0.1'

# 'messages' dictionary stores messages for users. The keys are usernames, and the values are lists of messages.
messages = defaultdict(list)

def handle_client(client, address):
    """
    Handles communication with a single client in a dedicated thread.
    :param client: The client socket.
    :param address: The address of the client.
    """
    print(f"Connected with {address}")
    client.send('You are connected to the server.'.encode('ascii'))

    # Variables to track the state of the current user and message reception.
    global messages
    receiving = False  # Flag to indicate if the server is expecting message content.
    currentUser = None  # The username of the client.
    toUser = None  # The recipient of the message being composed.

    while True:
        try:
            message: str = client.recv(1024).decode('ascii')
            if not message:
                break  # Break out of the loop if no message is received.
            print(f"Received from {address}: {message}")

            if currentUser is None:
                # The first command must be LOGIN with a username.
                if message.startswith("LOGIN") and len(message.split()) == 2:
                    currentUser = message.split()[1]
                    # Send the number of unread messages to the user upon login.
                    msg = str(len(messages[currentUser])) + '\n'
                    client.send(msg.encode('ascii'))
                else:
                    print("Please issue LOGIN command first")
            else:
                if receiving:
                    # If in receiving mode, store the incoming message for the recipient.
                    messages[toUser].append({'from': currentUser, 'message': message})
                    receiving = False
                    client.send("MESSAGE SENT\n".encode('ascii'))
                else:
                    # Process commands: READ, COMPOSE, or EXIT.
                    if message.startswith("READ"):
                        if len(messages[currentUser]) > 0:
                            response = messages[currentUser].pop(0)
                            user = response['from'] + "\n"
                            client.send(user.encode('ascii'))
                            msg = response['message'] + '\n'
                            client.send(msg.encode('ascii'))
                        else:
                            client.send("READ ERROR\n".encode('ascii'))
                    elif message.startswith("COMPOSE") and len(message.split()) == 2:
                        toUser = message.split()[1]
                        receiving = True
                    elif message.startswith("EXIT"):
                        client.send("Exiting...".encode('ascii'))
                        break
                    else:
                        client.send("Unknown command".encode('ascii'))
        except Exception as e:
            print(f"Error with {address}: {e}")
            break

    print(f"Connection closed with {address}")
    client.close()

def main(port):
    """
    The main function that sets up the server.
    :param port: The port number on which the server will listen for incoming connections.
    """
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        try:
            server_socket.bind((HOST, port))
        except socket.error as exception:
            print(f'Bind failed. Error Code : {str(exception[0])} Message {exception[1]}')
            sys.exit()

        server_socket.listen()
        print(f"Server listening on {HOST}:{port}")

        while True:
            client, address = server_socket.accept()
            thread = threading.Thread(target=handle_client, args=(client, address))
            thread.start()

if __name__ == '__main__':
    args = sys.argv[1:]
    if len(args) == 1 and args[0].isdigit():
        main(int(args[0]))
    else:
        print("Correct usage: script.py <PORT>")

